### PR TITLE
sock/dtls: add getter/setter for the remote UDP endpoint of sock_dtls_session_t

### LINKED ISSUE #15615: pkg/tinydtls: Question about the design of sock_dtls_session_t
#### Description
I have a few questsions regarding the implementation of `sock_dtls_session_t` of the tinyDTLS implementation. This is the struct for tinyDTLS:
```
struct sock_dtls_session {
 sock_udp_ep_t ep; 
 session_t dtls_session;
};
```
It consists of a UDP endpoint and a session_t struct, which itself is a session object of tinyDTLS. But if you look at `session_t`
```
typedef struct {
 unsigned char size;
 ipv6_addr_t addr;
 unsigned short port;
 int ifindex;
} session_t;
```
you can see, that it is just an UDP endpoint only represented differently. So...a `sock_dtls_session_t` for tinyDTLS is a UDP endpoint...twice. 

If I run into problems with the DTLS API I came to the conclusion that I am mostly bogging around the problem that sometimes one of the endpoints is not properly set. Additionally, a session object costs really much memory in its current state. So storing this struct should be highly avoided IMHO. 

Is there a reason for this double UDP endpoint design? As far as I can tell the `session_t` should be sufficient?

I can imagine that the idea behind this structure was, that you can easily set the session endpoint from a `sock_udp_ep_t` and that the GNRC udp endpoint is available for every DTLS implementation of `sock_dtls_session_t`. But I would always prefer an official way to set and retrieve the UDP endpoint from a session, so that I don't care how the endpoint is stored internally:
```
sock_dtls_session_get_udp_ep(...)
sock_dtls_session_set_udp_ep(...)
```

### PR DESCRIPTION
**Pull request by janosbrodbeck**
### Contribution description
This PR adds two new functions to the DTLS sock API: `sock_dtls_session_get_udp_ep()` and `sock_dtls_session_set_udp_ep()`.
Those functions add the ability to add or get the remote UDP endpoint from a session object (`sock_dtls_session_t`). 

__Necessity of the setter:__
When using `sock_dtls_session_init()` to initiate a handshake there is no need to set the remote endpoint because the function has the endpoint as parameter and sets the endpoint itself. But there is another legitimate way to initiate a handshake: by using `sock_dtls_send()`/`sock_dtls_send_aux()`. 

These functions start a handshake when no session exists and the timeout is > 0. But for that the endpoint needs to be set since it is not done by the function (since there is no endpoint parameter and it would not make sense to add it to that function).

From the documentation: 
```
timeout             Handshake timeout in microseconds.
 *                      If `timeout > 0`, will start a new handshake if no
 *                      session exists yet. The function will block until
 *                      handshake completed or timed out.
```

### Testing procedure
Implementation in tinyDTLS can be tested with the test [pkg_tinydtls_sock_async](https://github.com/RIOT-OS/RIOT/tree/master/tests/pkg_tinydtls_sock_async) together with this patch:
```
diff --git a/tests/pkg_tinydtls_sock_async/dtls-client.c b/tests/pkg_tinydtls_sock_async/dtls-client.c
index 2aefc75856..b03ac19ed7 100644
--- a/tests/pkg_tinydtls_sock_async/dtls-client.c
+++ b/tests/pkg_tinydtls_sock_async/dtls-client.c
@@ -200,6 +200,22 @@ static int client_send(char *addr_str, char *data)
         return -1;
     }
 
+    sock_dtls_session_t test_session;
+    sock_dtls_session_set_udp_ep(&test_session, &remote);
+
+    sock_udp_ep_t test_ep;
+    sock_dtls_session_get_udp_ep(&test_session, &test_ep);
+
+    printf("\n\nAddress: ");
+    for (uint8_t i=0; i < sizeof(ipv6_addr_t); i++) {
+        printf("%02x:", test_ep.addr.ipv6[i]);
+    }
+    puts("");
+    printf("Port: %d\n", test_ep.port);
+    printf("Family: %d\n", test_ep.family);
+    printf("Interface: %d\n", test_ep.netif);
+    puts("\n");
+
     if (sock_dtls_create(&_dtls_sock, &_udp_sock,
                          SOCK_DTLS_CLIENT_TAG,
                          SOCK_DTLS_1_2, SOCK_DTLS_CLIENT) < 0) {

```

To run it: 
* Flash and run on one device (e.g. native)
* Execute `dtlsc {ip} {some data}`
The printed UDP endpoint should be the valid remote endpoint. (Send fails because we have no server device)

### Issues/PRs references
Fixes #15615

### CODE REVIEW

**Comment by janosbrodbeck** [comment]
Could someone please link this PR to the issue #15615? When this is done everything from the issue should be addressed.

**Comment by pokgak** [comment]
Would it make sense to also replace `_session_to_ep` and `_ep_to_session` with these functions too because they are doing the same thing?

**Comment by miri64** [comment]
> Could someone please link this PR to the issue #15615? When this is done everything from the issue should be addressed.

Done. As described in the template, you need to use the right wording for Github to recognize it. I changed your issue reference to "Fixes #15615" in op :-).

File: sys/include/net/sock/dtls.h [hunk 1]
```diff
@@ -664,6 +664,31 @@ int sock_dtls_session_init(sock_dtls_t *sock, const sock_udp_ep_t *ep,
...
+/**
+ * @brief Set the remote UDP endpoint from a session.
+ *
+ * @pre `(session != NULL) && (ep != NULL)`
+ *
+ * @param[in]   session   DTLS session
+ * @param[in]   ep        UDP endpoint
+ *
+ * @note Function should only be needed when doing a blocking handshake with
+ *       @ref sock_dtls_send() to set the remote endpoint.
```
Comment (miri64) [inline sys/include/net/sock/dtls.h:687]: Maybe add some example usage code here?

**Comment by janosbrodbeck** [comment]
> Would it make sense to also replace `_session_to_ep` and `_ep_to_session` with these functions too because they are doing the same thing?

True. And we can also tick ` Add general way to set/get sock_dtls_session_t from a sock_udp_ep_t and vice-versa` in #10897 if this is done.

I'll add a commit to replace this helper functions.

**Comment by janosbrodbeck** [comment]
> Maybe add some example usage code here?

An example doesn't really fit there in my opinion. I've added an exemplary usage of blocking `sock_dtls_send()` to the note section of the two send functions. Theoretically you could have also put it to `sock_dtls_session_init()` as well. 
Is that sufficient/fitting for you?

> Would it make sense to also replace _session_to_ep and _ep_to_session with these functions too because they are doing the same thing?

I've looked at it again and...I don't think it is worth the hassle. Some functions call the helper functions without having a `sock_dtls_session_t` object. E.g. the `_write()` function. If I would remove the helper function I would have to add a `sock_dtls_session_t` object at these spots just to call the new functions instead.

File: sys/include/net/sock/dtls.h [hunk 2]
```diff
@@ -871,6 +871,14 @@ static inline ssize_t sock_dtls_recv_buf(sock_dtls_t *sock,
  * @note    When blocking, we will need an extra thread to call
  *          @ref sock_dtls_recv() function to handle the incoming handshake
  *          messages.
+ *          An exemplary send with a blocking handshake is:
```
Comment (miri64) [inline sys/include/net/sock/dtls.h:874]:
```suggestion
An example for a blocking handshake is:
```

File: sys/include/net/sock/dtls.h [hunk 3]
```diff
@@ -907,6 +915,14 @@ ssize_t sock_dtls_send_aux(sock_dtls_t *sock, sock_dtls_session_t *remote,
...
```
Comment (miri64) [inline sys/include/net/sock/dtls.h:918]: Here too.

File: sys/include/net/sock/dtls.h (hunk 2 above)
Comment (miri64) [reply sys/include/net/sock/dtls.h:874]: I think that example is enough to be mentioned once in `sock_dtls_send()` btw.

**Review by miri64 — CHANGES_REQUESTED** [review]

File: pkg/tinydtls/contrib/sock_dtls.c [hunk 4]
```diff
@@ -345,6 +345,24 @@ void sock_dtls_session_destroy(sock_dtls_t *sock, sock_dtls_session_t *remote)
     dtls_close(sock->dtls_ctx, &remote->dtls_session);
 }
 
+void sock_dtls_session_get_udp_ep(const sock_dtls_session_t *session,
+                                  sock_udp_ep_t *ep)
+{
+    assert(session);
+    assert(ep);
+
+    _session_to_ep(&session->dtls_session, ep);
```
Comment (miri64) [inline pkg/tinydtls/contrib/sock_dtls.c:354]: Would it make sense to replace the calls of this function with `sock_dtls_session_get_udp_ep()` in general?

**Comment by janosbrodbeck** [comment]
Applied the change request and removed the example from the `sock_dtls_send_aux()`.

**Comment by janosbrodbeck** [comment]
Squashed.

**Review by miri64 — APPROVED** [review]
Tested as suggested and it works :-). ACK!

**Pull request merged**